File: LeetCode/Graph/DFS/Python/Leetcode1161-parsing-a-boolean-expression.py
class Solution:
    def parseBoolExpr(self, s: str) -> bool:
        n, i = len(s), 0
        def dfs():
            nonlocal i
            if s[i] == 't': 
                i += 1
                return True
            if s[i] == 'f':
                i += 1
                return False
            # op
            op = s[i]
            i += 2
            res = True
            if op == '|': res = False
            while s[i] != ')':
                if s[i] == ',': i += 1
                t = dfs()
                if op == '|': res |= t
                else: res &= t
            i += 1
            if op == '!': res = not res
            return res
        return dfs()

File: LeetCode/Graph/DFS/JS/Leetcode1161-parsing-a-boolean-expression.js
var parseBoolExpr = function (s) {
  let i = 0;
  let dfs = () => {
    if (s[i] == 't') {
      i++;
      return true;
    }
    if (s[i] == 'f') {
      i++;
      return false;
    }

    // op
    const op = s[i];
    i += 2;
    let res = true;
    if (op == '|') res = false;
    while (s[i] != ')') {
      if (s[i] == ',') i++;
      const t = dfs();
      if (op == '|') res ||= t;
      else res &= t;
    }
    i++;
    if (op == '!') res = !res;
    return res;
  }
  return dfs();
};